generate/src/contract/deployment.rs

```diff
@@ -99,41 +98,49 @@ fn expand_deploy(cx: &Context) -> Result<TokenStream> {
         None => (quote! {}, quote! {()}),
     };
 
-    let lib_params: Vec<_> = cx
-        .artifact
-        .bytecode
-        .undefined_libraries()
-        .map(|name| Param {
-            name: name.to_snake_case(),
-            kind: ParamType::Address,
-        })
-        .collect();
-    let lib_input = methods::expand_inputs(&lib_params)?;
-
-    let link = if !lib_params.is_empty() {
-        let link_libraries = cx
-            .artifact
-            .bytecode
-            .undefined_libraries()
-            .zip(lib_params.iter())
-            .map(|(name, lib_param)| {
-                let name = Literal::string(&name);
-                let address = util::ident(&lib_param.name);
+    let libs: Vec<_> = cx.artifact.bytecode.undefined_libraries().collect();
+    let (lib_struct, lib_input, link) = if !libs.is_empty() {
+        let lib_struct = {
+            let lib_struct_fields = libs.iter().map(|name| {
+                let doc = util::expand_doc(&format!("Address of the `{}` library.", name));
+                let field = util::safe_ident(&name.to_snake_case());
+
+                quote! { #doc pub #field: self::ethcontract::Address }
+            });
+
+            quote! {
+                /// Undefinied libraries in the contract bytecode that are
+                /// required for linking in order to deploy.
+                pub struct Libraries {
+                    #( #lib_struct_fields, )*
+                }
+            }
+        };
+
+        let link = {
+            let link_libraries = libs.iter().map(|name| {
+                let name_lit = Literal::string(&name);
+                let field = util::safe_ident(&name.to_snake_case());
```

Comment:
nit: since this logic is shared with the loop above, how about making the one above return a pair of `(lib_struct_fields, link_libraries)`?

Reply:
Good point!